src/mlmc/quantity_concept.py

```diff
@@ -121,120 +373,594 @@ def mult_op(x, y):
         return self._reduction_op([self], cmult_op)
 
-def make_root_quantity(storage: Storage, q_specs : List[QSpec])
-    i_range = 0
-    dict_items = []
-    for q_spec in q_specs:
-        time_frames = []
-        for t in q_spec.times:
-            field_values = []
-            for loc in q_spec.locations:
-                q_range = QRange(storage)
-                array = Array.from_range(q_spec.shape, q_range)
-                q_range.set(i_range, array._size)
-                i_range += array._size
-                field_values.append( (loc, array) )
-            time_frames.append( (t, Field(field_values)) )
-        dict_items = (q_spec.name, TimeSeries(time_frames))
-    return Dict(dict_items)
-
-
-
-class QRange:
-    def __init__(self, storage):
-        self._storage = storage
+    @staticmethod
+    def _process_mask(x, y, operator, level_id):
+        """
+        Create samples mask
+        All values for sample must meet given condition, if any value doesn't meet the condition,
+        whole sample is eliminated
+        :param x: Quantity
+        :param y: Quantity or int, float
+        :param operator: operator module function
+        :param level_id: int, level identifier
+        :return: np.ndarray of bools
+        """
+        # Zero level - use just fine samples
+        if level_id == 0:
+            if isinstance(y, int) or isinstance(y, float):
+                mask = operator(x[..., 0], y)  # y is int or float
+            else:
+                mask = operator(x[..., 0], y[..., 0])  # y is from other quantity
+            return mask.all(axis=tuple(range(mask.ndim - 1)))
+
+        mask = operator(x, y)
+        return mask.all(axis=tuple(range(mask.ndim - 2))).all(axis=1)
+
+    def _mask_quantity(self, other, op):
+        """
+        Create quantity that represent bool mask
+        :param other: number or Quantity
+        :param op: operation
+        :return: Quantity
+        """
+        bool_type = BoolType()
+        new_qtype = copy.deepcopy(self.qtype)
+        new_qtype.replace_scalar(bool_type)
 
-    def set(self, range_begin, size):
-        self.begin = range_begin
-        self._size = size
-        self.end = range_begin + size
+        if isinstance(other, (float, int)):
+            if not isinstance(self.qtype.base_qtype(), ScalarType):
+                raise TypeError("Quantity has base qtype {}. "
+                                "Quantities with base qtype ScalarType are the only ones that support comparison".
+                                format(self.qtype.base_qtype()))
+
+            return Quantity(quantity_type=new_qtype, input_quantities=[self], operation=op)
+        elif isinstance(other, Quantity):
+            if self.qtype.size() != other.qtype.size():
+                raise Exception("Quantities don't have have the same size")
+            if not isinstance(self.qtype.base_qtype(), ScalarType) or not isinstance(other.qtype.base_qtype(), ScalarType):
```

Comment:
If we do not check this in other operations we should test it here neither. Or we should test it in the Quantity constructor and only suppress it when applying the mask.

Reply:
Same for the previous check. Either implement such checks systematically or not at all.

Reply:
Possibly move both checks into _reuce_op and use it here.

Reply:
I think `_reduction_op` is not suitable for this. We should discuss it.